src/bucket/BucketIndex.cpp

```diff
@@ -0,0 +1,339 @@
+    }
+
+    CLOG_DEBUG(Bucket, "Indexed {} positions in {}", mKeysToOffset.size(),
+               filename.filename());
+    ZoneValue(static_cast<int64_t>(mKeysToOffset.size()));
+}
+
+// Returns true if the key is not contained within the given IndexEntry.
+// Range index: check if key is outside range of indexEntry
+// Individual index: check if key does not match indexEntry key
+template <class IndexEntryT>
+static bool
+keyNotInIndexEntry(LedgerKey const& key, IndexEntryT const& indexEntry)
+{
+    if constexpr (std::is_same<IndexEntryT, BucketIndex::RangeEntry>::value)
+    {
+        return key < indexEntry.lowerBound || indexEntry.upperBound < key;
+    }
+    else
+    {
+        return !(key == indexEntry);
+    }
+}
+
+// std::lower_bound predicate. Returns true if index comes "before" key and does
+// not contain it
+// If key is too small for indexEntry bounds: return false
+// If key is contained within indexEntry bounds: return false
+// If key is too large for indexEntry bounds: return true
+template <class IndexEntryT>
+static bool
+lower_bound_pred(IndexEntryT const& indexEntry, LedgerKey const& key)
+{
+    if constexpr (std::is_same<IndexEntryT,
+                               BucketIndex::RangeIndex::value_type>::value)
+    {
+        return indexEntry.first.upperBound < key;
+    }
+    else
+    {
+        return indexEntry.first < key;
+    }
+}
+
+// std::upper_bound predicate. Returns true if key comes "before" and is not
+// contained within the indexEntry.
+// If key is too small for indexEntry bounds: return true
+// If key is contained within indexEntry bounds: return false
+// If key is too large for indexEntry bounds: return false
+template <class IndexEntryT>
+static bool
+upper_bound_pred(LedgerKey const& key, IndexEntryT const& indexEntry)
+{
+    if constexpr (std::is_same<IndexEntryT,
+                               BucketIndex::RangeIndex::value_type>::value)
+    {
+        return key < indexEntry.first.lowerBound;
+    }
+    else
+    {
+        return key < indexEntry.first;
+    }
+}
+
+std::unique_ptr<BucketIndex const>
+BucketIndex::createIndex(BucketManager const& bm,
+                         std::filesystem::path const& filename)
+{
+    ZoneScoped;
+    auto const& cfg = bm.getConfig();
+    releaseAssertOrThrow(cfg.EXPERIMENTAL_BUCKETLIST_DB);
+    releaseAssertOrThrow(!filename.empty());
+
+    auto pageSizeExp = cfg.EXPERIMENTAL_BUCKETLIST_DB_INDEX_PAGE_SIZE_EXPONENT;
+    releaseAssertOrThrow(pageSizeExp < 32);
+    auto pageSize = pageSizeExp == 0 ? 0 : 1UL << pageSizeExp;
+
+    // Conver to bytes
+    auto cutoff = cfg.EXPERIMENTAL_BUCKETLIST_DB_INDEX_CUTOFF * 1000000;
+    if (pageSize == 0 || fs::size(filename.string()) < cutoff)
+    {
+        CLOG_INFO(Bucket,
+                  "BucketIndex::createIndex() indexing individual keys in "
+                  "bucket {}",
+                  filename);
+        return std::unique_ptr<BucketIndexImpl<IndividualIndex> const>(
```

Comment:
I think it may be incorrect to return a unique ptr like this if indexing was not completed for whatever reason (both variants)? In particular the index is partial when aborting on shutdown.
In other places, the assumption is that if not null then the index is valid

Reply:
I agree that partial indexes shouldn't be returned. I initially had the index constructor throw if a shutdown was detected, but error messages were messy since the exception would not get caught. I'd like to keep the unique_ptr return interface, so maybe I'll put this throw back but add a try-catch in `createIndex`. If the `BucketIndex` constructor thows, `createIndex` can return `nullptr` so we still exit early but don't have an invalid index floating around.